Noviny/remove pages directory

Noviny opened the pull request:
This bug keeps slowing down dev speed so I made a quick fix ❤️

Blasz commented:
Couple of questions for context,

1. What was pages directory used for?
2. How was it slowing down dev speed?

Blasz commented on packages/website/src/bin/page-generator/test.js:
```diff
@@ -30,35 +32,42 @@ async function runGeneratePages({
+  const resolvedPackageRoot = packageRoot || (await createTempDir());
+
+  const tempDefaultPagesPath = path.join(resolvedPackageRoot, 'default-pages');
+  await fs.copy(defaultPagesPath, tempDefaultPagesPath);
```
Why do we need to copy the default-pages across, doesn't `generatePages` do that itself?

Noviny replied:
This copies the default pages into the test directory which we generate pages in - if they are not there, generatePages fails because it can't copy them across.

Blasz commented on packages/website/src/bin/page-generator/test.js:
```diff
@@ -30,35 +32,42 @@ async function runGeneratePages({
 
   describe('package pages generation', () => {
-    it('creates a home page per package', () => {
+    it('creates a home page per package', async () => {
```
Is this necessary?

Noviny replied:
No

Noviny commented:
> What was pages directory used for?

`pages` is the place next runs out of - currently we have a mix of generated pages and 'static' pages that were in git history. This meant when cleaning up the generated pages, we removed pages we were about to create. THIS meant if we changed what we were generating, you would have pages that next would build that you weren't intending.

We've had several instances where as people working on the project, we had outdated pages generated that had webpack errors that broke the build - it's unexpected why things are breaking so is hard to debug the first time.

For consumers, they can get into a state by changing their docs config where they cannot build until they rm -rf node_modules.

The solution is `pages` is entirely generated at runtime, and we can safely remove the entire thing on startup. We move the always-pages into default-pages, and copy that across to kick off the whole process

> How was it slowing down dev speed?

See above for annoying repeated bugs that took a while to debug (particularly if you hadn't encountered them before.

Blasz commented on packages/website/src/bin/page-generator/index.js:
```diff
@@ -315,22 +314,20 @@ const generateProjectDocsPages = (docsInfo, generatorConfig, name, urlPath) => {
+
+const addBasePages = async (packageRoot, pagesPath) => {
+  const defaultPagesPath = path.join(packageRoot, 'default-pages');
+  await fs.copy(defaultPagesPath, pagesPath);
```
Can we add some tests surrounding this functionality? Do we have any existing tests that test the presence of the default pages?

Noviny replied:
Every test in generatePages fails if this doesn't work.

Blasz replied:
Right, that's not equivalent to testing this functionality though. Is the fail criteria the fact that the 'pages' dir doesn't exist? If so, changing the code to `fs.mkdir('pages')` wouldn't break the other tests but would break functionality.

I think adding a test called 'copies default pages over to pages' would be good for documentation as well.

dpisani-atl replied:
Perhaps a new test that specifically asserts the new behaviour would be good as well.

Noviny replied:
Added a test

Blasz commented on packages/website/src/bin/page-generator/test.js:
```diff
@@ -408,12 +419,21 @@ describe('Generate pages', () => {
+    expect(fs.existsSync(path.join(pagesPath, '_document.tsx'))).toEqual(true);
+    expect(fs.existsSync(path.join(pagesPath, '_error.tsx'))).toEqual(true);
+    expect(fs.existsSync(path.join(pagesPath, 'index.tsx'))).toEqual(true);
+    expect(fs.existsSync(path.join(pagesPath, 'packages.tsx'))).toEqual(true);
```
`toMatchSnapshot`? :)
